File: ios/src/main/kotlin/ui/AboutViewController.kt
import kotlinx.cinterop.*
import platform.UIKit.*
import platform.MapKit.*
import platform.Foundation.*
import platform.CoreGraphics.*
import platform.CoreLocation.*

@ExportObjCClass
class AboutViewController : UIViewController, MKMapViewDelegateProtocol {
    companion object {
        val SHOW_BADGE_KEY = "showBadge"
    }

    @ObjCOutlet lateinit var map: MKMapView
    @ObjCOutlet lateinit var badge: UIView

    constructor(aDecoder: NSCoder) : super(aDecoder)
    override fun initWithCoder(aDecoder: NSCoder) = initBy(AboutViewController(aDecoder))

    override fun viewDidLoad() {
        map.applyCornerRadius(6.0)
        
        map.delegate = this

        val location = CLLocation(37.805423, longitude = -122.401123)
        map.setRegion(MKCoordinateRegionMakeWithDistance(location.coordinate(), 800.0, 800.0), animated = false)

        val confPlace = MKPointAnnotation().apply {
            setCoordinate(location.coordinate())
            setTitle("Pier 27")
        }

        badge.hidden = !NSUserDefaults.standardUserDefaults.boolForKey(SHOW_BADGE_KEY)

        map.addAnnotation(confPlace)
    }
    
    @ObjCAction 
    fun onMapClick(sender: ObjCObject?) {
        val place = map.annotations.firstObject.uncheckedCast<MKPointAnnotation>()
        val mapItem = MKMapItem(MKPlacemark(place.coordinate(), addressDictionary = null))
        mapItem.name = place.title()?.toString()
        mapItem.openInMapsWithLaunchOptions(null)
    }
    
    @ObjCAction 
    fun onTwitterClick(sender: ObjCObject?) {
        openUrl("https://twitter.com/kotlinconf")
    }
    
    @ObjCAction 
    fun onWebsiteClick(sender: ObjCObject?) {
        openUrl("https://www.kotlinconf.com/")
    }
    
    override fun mapView(mapView: MKMapView, viewForAnnotation: MKAnnotationProtocol): MKAnnotationView? {
        val identifier = "pin"
        
        val dequeuedView = map.dequeueReusableAnnotationViewWithIdentifier(identifier)
        if (dequeuedView is MKPinAnnotationView) {
            dequeuedView.annotation = viewForAnnotation
            return dequeuedView
        }

        return MKPinAnnotationView(viewForAnnotation, reuseIdentifier = identifier).apply {
            canShowCallout = true
            calloutOffset = CGPointMake(-5.0, 5.0)
            rightCalloutAccessoryView = UIButton.buttonWithType(UIButtonTypeDetailDisclosure)
        }
    }
}

File: ios/konfswift/ui/AboutViewController.swift
import UIKit
import MapKit
import AddressBook
import Contacts

class AboutViewController : UIViewController, MKMapViewDelegate {
    @IBOutlet private weak var map: MKMapView!

    override func viewDidLoad() {
        map.delegate = self

        let location = CLLocation(latitude: 37.805423, longitude: -122.401123)
        map.setRegion(MKCoordinateRegionMakeWithDistance(location.coordinate, 400, 400), animated: false)

        let confPlace = Place(title: "Pier 27", subtitle: "The Embarcadero, San Francisco", coordinate: location.coordinate)
        map.addAnnotation(confPlace)
    }

    func mapView(_ mapView: MKMapView, viewFor annotation: MKAnnotation) -> MKAnnotationView? {
        guard annotation is Place else { return nil }

        let identifier = "pin"
        let view: MKPinAnnotationView
        if let dequeuedView = map.dequeueReusableAnnotationView(withIdentifier: identifier) as? MKPinAnnotationView {
            dequeuedView.annotation = annotation
            view = dequeuedView
        } else {
            view = MKPinAnnotationView(annotation: annotation, reuseIdentifier: identifier)
            view.canShowCallout = true
            view.calloutOffset = CGPoint(x: -5, y: 5)
            view.rightCalloutAccessoryView = UIButton(type: .detailDisclosure)
        }

        return view
    }

    func mapView(_ mapView: MKMapView, annotationView view: MKAnnotationView, calloutAccessoryControlTapped control: UIControl) {
        guard let place = view.annotation as? Place else { return }
        let mapItem = MKMapItem(placemark: MKPlacemark(coordinate: place.coordinate, addressDictionary: nil))
        mapItem.name = place.title
        _ = mapItem.openInMaps(launchOptions: nil)
    }
}

fileprivate class Place: NSObject, MKAnnotation {
    let title: String?
    let subtitle: String?
    let coordinate: CLLocationCoordinate2D

    init(title: String, subtitle: String, coordinate: CLLocationCoordinate2D) {
        self.title = title
        self.subtitle = subtitle
        self.coordinate = coordinate

        super.init()
    }
}
